# SPDX-License-Identifier: Apache-2.0
# Downstream Contextlib2 21.6.0 stub-only compatibility adjustment.
# Declare existing concrete exit methods and actual coroutine signatures.
# Runtime implementation, tests and original stubtest allowlist are unchanged.
--- a/contextlib2/__init__.pyi
+++ b/contextlib2/__init__.pyi
@@ -41,4 +41,5 @@
 class _GeneratorContextManager(ContextManager[_T_co]):
-    def __call__(self, func: _F) -> _F: ...
+    def __call__(self, func: _F) -> _F: ...
+    def __exit__(self, type: Optional[Type[BaseException]], value: Optional[BaseException], traceback: Optional[TracebackType]) -> bool: ...
-
+
 # type ignore to deal with incomplete ParamSpec support in mypy
@@ -55,4 +56,5 @@
 class closing(ContextManager[_SupportsCloseT]):
-    def __init__(self, thing: _SupportsCloseT) -> None: ...
+    def __init__(self, thing: _SupportsCloseT) -> None: ...
+    def __exit__(self, *exc_info: Any) -> None: ...
-
+
 if True:
@@ -61,5 +63,6 @@
     _SupportsAcloseT = TypeVar("_SupportsAcloseT", bound=_SupportsAclose)
     class aclosing(AsyncContextManager[_SupportsAcloseT]):
-        def __init__(self, thing: _SupportsAcloseT) -> None: ...
+        def __init__(self, thing: _SupportsAcloseT) -> None: ...
+        async def __aexit__(self, *exc_info: Any) -> None: ...
     _AF = TypeVar("_AF", bound=Callable[..., Awaitable[Any]])
     class AsyncContextDecorator:
@@ -74,4 +77,5 @@
 class redirect_stdout(ContextManager[_T_io]):
-    def __init__(self, new_target: _T_io) -> None: ...
+    def __init__(self, new_target: _T_io) -> None: ...
+    def __exit__(self, exctype: Optional[Type[BaseException]], excinst: Optional[BaseException], exctb: Optional[TracebackType]) -> None: ...
-
+
 class redirect_stderr(ContextManager[_T_io]):
@@ -77,4 +81,5 @@
 class redirect_stderr(ContextManager[_T_io]):
-    def __init__(self, new_target: _T_io) -> None: ...
+    def __init__(self, new_target: _T_io) -> None: ...
+    def __exit__(self, exctype: Optional[Type[BaseException]], excinst: Optional[BaseException], exctb: Optional[TracebackType]) -> None: ...
-
+
 class ContextDecorator:
@@ -115,10 +120,10 @@
         def aclose(self) -> Awaitable[None]: ...
-        def __aenter__(self: _S) -> Awaitable[_S]: ...
-        def __aexit__(
+        async def __aenter__(self: _S) -> _S: ...
+        async def __aexit__(
             self,
             __exc_type: Optional[Type[BaseException]],
             __exc_value: Optional[BaseException],
             __traceback: Optional[TracebackType],
-        ) -> Awaitable[bool]: ...
+        ) -> bool: ...
-
+
 if True:
